From 3151ab36dcbf8fd43d3385ca2ed13a1e50422669 Mon Sep 17 00:00:00 2001
From: "linfang.zhao" <linfang.zhao@amlogic.com>
Date: Tue, 21 Nov 2023 03:22:21 +0000
Subject: [PATCH 7/9] drm: switch fail between ubuntu ui and fb console [1/1]

PD#SWPL-132042

Problem:
fb phy addr not flush when switch to fb console and fbdev commit

Solution:
flush fb phy addr

Verify:
T7C

Test:
DRM-OSD-112

Change-Id: I3c471964d939399b79826f26d6b98fa18737d404
Signed-off-by: linfang.zhao <linfang.zhao@amlogic.com>
---
 drivers/drm/meson_fbdev.c              | 5 ++++-
 drivers/drm/meson_plane.c              | 1 +
 drivers/drm/meson_plane.h              | 1 +
 drivers/drm/meson_vpu_pipeline.h       | 2 ++
 drivers/drm/vpu-hw/meson_vpu_osd_mif.c | 6 ++++--
 5 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/drivers/drm/meson_fbdev.c b/drivers/drm/meson_fbdev.c
index d15b0cf4d..374c56ad6 100644
--- a/drivers/drm/meson_fbdev.c
+++ b/drivers/drm/meson_fbdev.c
@@ -465,6 +465,7 @@ static int am_meson_drm_fb_pan_display(struct fb_var_screeninfo *var,
 	struct drm_device *dev = fb_helper->dev;
 	struct drm_atomic_state *state;
 	struct drm_plane_state *plane_state;
+	struct am_meson_plane_state *am_plane_state;
 	struct drm_plane *plane = fbdev->plane;
 	struct drm_mode_set *mode_set;
 	int hdisplay, vdisplay;
@@ -533,10 +534,12 @@ retry:
 		plane_state->crtc_h);
 
 	state->legacy_cursor_update = true;
+	am_plane_state = to_am_meson_plane_state(plane_state);
+	am_plane_state->fbdev_commit = true;
 	ret = drm_atomic_commit(state);
 	if (ret != 0)
 		goto fail;
-
+	am_plane_state->fbdev_commit = false;
 	info->var.xoffset = var->xoffset;
 	info->var.yoffset = var->yoffset;
 
diff --git a/drivers/drm/meson_plane.c b/drivers/drm/meson_plane.c
index 2f2379350..fc1823899 100644
--- a/drivers/drm/meson_plane.c
+++ b/drivers/drm/meson_plane.c
@@ -1234,6 +1234,7 @@ static int meson_plane_atomic_check(struct drm_plane *plane,
 
 	plane_state = to_am_meson_plane_state(state);
 	plane_info->sec_en = plane_state->sec_en;
+	plane_info->fbdev_commit = plane_state->fbdev_commit;
 
 	DRM_DEBUG("OSD PLANE index=%d, zorder=%d, premult= %d, alpha = %d, phy = %llx\n",
 		  plane_info->plane_index, plane_info->zorder,
diff --git a/drivers/drm/meson_plane.h b/drivers/drm/meson_plane.h
index a7a150166..ba19cf181 100644
--- a/drivers/drm/meson_plane.h
+++ b/drivers/drm/meson_plane.h
@@ -24,6 +24,7 @@
 struct am_meson_plane_state {
 	struct drm_plane_state base;
 	u32 sec_en;
+	bool fbdev_commit;
 };
 
 enum meson_max_fb_enum {
diff --git a/drivers/drm/meson_vpu_pipeline.h b/drivers/drm/meson_vpu_pipeline.h
index bf95ea447..1e1286332 100644
--- a/drivers/drm/meson_vpu_pipeline.h
+++ b/drivers/drm/meson_vpu_pipeline.h
@@ -192,6 +192,7 @@ struct meson_vpu_osd_layer_info {
 	u32 status_changed;
 	int sec_en;
 	u32 *palette_arry;
+	bool fbdev_commit;
 };
 
 enum osd_mif_acc_mode {
@@ -250,6 +251,7 @@ struct meson_vpu_osd_state {
 	u32 read_ports;
 	int sec_en;
 	u32 *palette_arry;
+	bool fbdev_commit;
 };
 
 struct meson_vpu_video_layer_info {
diff --git a/drivers/drm/vpu-hw/meson_vpu_osd_mif.c b/drivers/drm/vpu-hw/meson_vpu_osd_mif.c
index 24826e97c..93e1b5330 100644
--- a/drivers/drm/vpu-hw/meson_vpu_osd_mif.c
+++ b/drivers/drm/vpu-hw/meson_vpu_osd_mif.c
@@ -1199,6 +1199,7 @@ static int osd_check_state(struct meson_vpu_block *vblk,
 	mvos->read_ports = plane_info->read_ports;
 	mvos->sec_en = plane_info->sec_en;
 	mvos->palette_arry = plane_info->palette_arry;
+	mvos->fbdev_commit = plane_info->fbdev_commit;
 
 	return 0;
 }
@@ -1360,8 +1361,9 @@ static void osd_set_state(struct meson_vpu_block *vblk,
 		hold_line = osd_hold_line;
 
 	flush_reg = osd_check_config(mvos, old_mvos);
-	MESON_DRM_BLOCK("flush_reg-%d index-%d\n", flush_reg, vblk->index);
-	if (!flush_reg) {
+	MESON_DRM_BLOCK("flush_reg-%d index-%d, fbdev_commit-%d\n",
+		flush_reg, vblk->index, mvos->fbdev_commit);
+	if (!mvos->fbdev_commit && !flush_reg) {
 		/*after v7 chips, always linear addr*/
 		if (osd->mif_acc_mode == LINEAR_MIF)
 			osd_mem_mode(vblk, reg_ops, reg, 1);
-- 
2.25.1

